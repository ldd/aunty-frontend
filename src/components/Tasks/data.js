const generateFakeData = () =>
  Array(31)
    .fill(true)
    .map(_ => Math.random() < 0.5);

export const prepareData = (data = []) =>
  data.reduce(
    (acc, n, i) =>
      i % 7 === 0
        ? [...acc, [n]]
        : [...acc.slice(0, acc.length - 1), acc[acc.length - 1].concat(n)],
    []
  );

const TASKS_URL = "http://localhost:4000/v1/tasks";

export const fetchData = async (
  setter = () => {},
  failer = () => {},
  url = TASKS_URL,
  options = {}
) => {
  const rawData = await fetch(url, {
    credentials: "include",
    ...options
  });
  if (rawData.ok) {
    const data = await rawData.json();
    setter(data);
  } else {
    failer();
  }
};

export const deleteTask = async (id = -1, setter, failer) =>
  fetchData(setter, failer, `${TASKS_URL}/${id}`, { method: "DELETE" });

export const fakeTasks = [
  {
    label: "Gym",
    description: "Go to the Gym",
    data: generateFakeData(),
    id: "id#Gym"
  },
  {
    label: "Read",
    description: "Read 10 pages",
    data: generateFakeData(),
    id: "id#Read"
  }
];
