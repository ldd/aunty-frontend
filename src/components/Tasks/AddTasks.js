import React, { useEffect, useState, useCallback } from "react";
import Heading from "../Heading";
import { addTask } from "./data";

const Field = ({ label, value, setValue, hint }) => {
  const onChange = e => setValue(e.target.value);
  return (
    <div className="field" style={{ textAlign: "left" }}>
      <label className="label">{label}</label>
      <div className="control has-icons-left has-icons-right">
        <input
          className="input is-success"
          type="text"
          value={value}
          onChange={onChange}
        />
        <span className="icon is-small is-left">
          <i className="fas fa-user" />
        </span>
        <span className="icon is-small is-right">
          <i className="fas fa-check" />
        </span>
      </div>
      <p className="help is-dark">{hint}</p>
    </div>
  );
};

const Control = ({ onSubmit, onCancel }) => (
  <div className="field is-grouped is-grouped-right">
    <div className="control">
      <div onClick={onCancel} className="button is-light">
        Cancel
      </div>
    </div>
    <div className="control">
      <div onClick={onSubmit} className="button is-primary">
        Submit
      </div>
    </div>
  </div>
);
const Form = ({ hideModal, onTaskAdded }) => {
  // field data
  const [label, setLabel] = useState("");
  const [description, setDescription] = useState("");
  const [frequency, setFrequency] = useState("");
  // for submitting the form
  const [payload, setPayload] = useState(null);
  const onSubmit = () => {
    setPayload({
      task_label: label,
      task_description: description,
      task_frequency: frequency
    });
  };
  useEffect(() => {
    if (payload !== null) {
      addTask(
        JSON.stringify(payload),
        (...args) => {
          hideModal();
          onTaskAdded(...args);
        },
        hideModal
      );
    }
  }, [payload, hideModal, onTaskAdded]);

  return (
    <div className="box">
      <Field
        label="Label"
        value={label}
        hint="example: read"
        setValue={setLabel}
      />
      <Field
        label="Description"
        value={description}
        hint="example: read at least 10 pages"
        setValue={setDescription}
      />
      <Field
        label="Frequency"
        value={frequency}
        hint="examples: monday, wednesday, friday | daily | weekends"
        setValue={setFrequency}
      />
      <Control onSubmit={onSubmit} onCancel={hideModal} />
    </div>
  );
};
const AddTasks = ({ onTaskAdded }) => {
  const [modal, setModal] = useState(false);
  const showModal = () => setModal(true);
  const hideModal = useCallback(() => setModal(false), []);
  return (
    <>
      <div
        onClick={showModal}
        className="button is-large is-success is-inverted"
        style={{ fontSize: "2.5rem" }}
      >
        +
      </div>
      <div className={`modal ${modal ? "is-active" : ""}`}>
        <div className="modal-background" />
        <div className="modal-content">
          <Heading title="Add Task" />
          <Form hideModal={hideModal} onTaskAdded={onTaskAdded} />
        </div>
        <button
          onClick={hideModal}
          className="modal-close is-large"
          aria-label="close"
        />
      </div>
    </>
  );
};

export default AddTasks;
