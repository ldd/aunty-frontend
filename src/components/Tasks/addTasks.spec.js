import React from "react";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { shallow, mount } from "enzyme";
import AddTasks, { Form, Field, Control } from "./AddTasks";
import * as data from "./data";

describe("addTasks", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    fetch.resetMocks();
  });

  it("renders snapshot", () => {
    const component = renderer.create(<AddTasks />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("should submit an empty form with no errors", () => {
    const wrapper = shallow(<AddTasks />);
    wrapper
      .find(Form)
      .dive()
      .find(Control)
      .prop("onSubmit")();
  });
  it("should use onTaskAdded prop", async () => {
    const mockedAddTask = jest.fn();

    const wrapper = mount(<AddTasks onTaskAdded={mockedAddTask} />);
    wrapper
      .find("input")
      .forEach(node =>
        node.simulate("change", { target: { value: "someValue" } })
      );

    fetch.mockResponse(JSON.stringify(data.fakeTasks));
    await act(async () => await wrapper.find(Control).invoke("onSubmit")());
    expect(mockedAddTask).toHaveBeenCalledWith(data.fakeTasks);

    const [, options] = fetch.mock.calls[0];
    expect(options.body).toContain("someValue");
  });

  it("should call addTask helper function", async () => {
    const mockedAddTask = jest.fn();
    jest.spyOn(data, "addTask").mockImplementation(mockedAddTask);

    const wrapper = mount(<AddTasks onTaskAdded={mockedAddTask} />);
    wrapper
      .find("input")
      .forEach(node =>
        node.simulate("change", { target: { value: "someValue" } })
      );

    await act(async () => await wrapper.find(Control).invoke("onSubmit")());
    expect(mockedAddTask).toHaveBeenCalled();
  });
});
