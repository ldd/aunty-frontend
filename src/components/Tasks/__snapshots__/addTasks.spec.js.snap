// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTasks renders snapshot 1`] = `
Array [
  <div
    className="button is-large is-success is-inverted"
    onClick={[Function]}
    style={
      Object {
        "fontSize": "2.5rem",
      }
    }
  >
    +
  </div>,
  <div
    className="modal "
  >
    <div
      className="modal-background"
    />
    <div
      className="modal-content"
    >
      <h1
        className="title"
      >
        Add Task
      </h1>
      <div
        className="box"
      >
        <div
          className="field"
          style={
            Object {
              "textAlign": "left",
            }
          }
        >
          <label
            className="label"
          >
            Label
          </label>
          <div
            className="control has-icons-left has-icons-right"
          >
            <input
              className="input is-success"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="icon is-small is-left"
            >
              <i
                className="fas fa-user"
              />
            </span>
            <span
              className="icon is-small is-right"
            >
              <i
                className="fas fa-check"
              />
            </span>
          </div>
          <p
            className="help is-dark"
          >
            example: read
          </p>
        </div>
        <div
          className="field"
          style={
            Object {
              "textAlign": "left",
            }
          }
        >
          <label
            className="label"
          >
            Description
          </label>
          <div
            className="control has-icons-left has-icons-right"
          >
            <input
              className="input is-success"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="icon is-small is-left"
            >
              <i
                className="fas fa-user"
              />
            </span>
            <span
              className="icon is-small is-right"
            >
              <i
                className="fas fa-check"
              />
            </span>
          </div>
          <p
            className="help is-dark"
          >
            example: read at least 10 pages
          </p>
        </div>
        <div
          className="field"
          style={
            Object {
              "textAlign": "left",
            }
          }
        >
          <label
            className="label"
          >
            Frequency
          </label>
          <div
            className="control has-icons-left has-icons-right"
          >
            <input
              className="input is-success"
              onChange={[Function]}
              type="text"
              value=""
            />
            <span
              className="icon is-small is-left"
            >
              <i
                className="fas fa-user"
              />
            </span>
            <span
              className="icon is-small is-right"
            >
              <i
                className="fas fa-check"
              />
            </span>
          </div>
          <p
            className="help is-dark"
          >
            examples: monday, wednesday, friday | daily | weekends
          </p>
        </div>
        <div
          className="field is-grouped is-grouped-right"
        >
          <div
            className="control"
          >
            <div
              className="button is-light"
              onClick={[Function]}
            >
              Cancel
            </div>
          </div>
          <div
            className="control"
          >
            <div
              className="button is-primary"
              onClick={[Function]}
            >
              Submit
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      aria-label="close"
      className="modal-close is-large"
      onClick={[Function]}
    />
  </div>,
]
`;
